tests/unit/sql_test.py

```diff
@@ -0,0 +1,302 @@
+        result.append(next(i).result().get_object_with_index(0))
+        result.append(next(i).result().get_object_with_index(0))
+
+        self.assertEqual(EXPECTED_ROWS, result)
+
+        # initiate the fetch request
+        future = next(i)
+
+        self.result.close()
+
+        with self.assertRaises(HazelcastSqlError) as cm:
+            future.result()
+
+        self.assertEqual(_SqlErrorCode.CANCELLED_BY_USER, cm.exception._code)
+
+    def set_fetch_response_with_error(self):
+        response = {"row_page": None, "error": _SqlError(_SqlErrorCode.PARSING, "expected", None)}
+        self.set_future_result_or_exception(response, sql_fetch_codec._REQUEST_MESSAGE_TYPE)
+
+    def set_fetch_error(self, error):
+        self.set_future_result_or_exception(error, sql_fetch_codec._REQUEST_MESSAGE_TYPE)
+
+    def set_close_error(self, error):
+        self.set_future_result_or_exception(error, sql_close_codec._REQUEST_MESSAGE_TYPE)
+
+    def set_close_response(self):
+        self.set_future_result_or_exception(None, sql_close_codec._REQUEST_MESSAGE_TYPE)
+
+    def set_execute_response_with_update_count(self):
+        self.set_execute_response(EXPECTED_UPDATE_COUNT, None, None, None)
+
+    def get_rows_from_blocking_iterator(self):
+        return [row.get_object_with_index(0) for row in self.result]
+
+    def get_rows_from_iterator(self):
+        result = []
+        for row_future in self.result.iterator().result():
+            try:
+                row = row_future.result()
+                result.append(row.get_object_with_index(0))
+            except StopIteration:
+                break
+        return result
+
+    def set_execute_response_with_rows(self, is_last=True):
+        self.set_execute_response(
+            -1,
+            [SqlColumnMetadata("name", SqlColumnType.VARCHAR, True, True)],
+            _SqlPage([SqlColumnType.VARCHAR], [EXPECTED_ROWS], is_last),
+            None,
+        )
+
+    def set_execute_response(self, update_count, row_metadata, row_page, error):
+        response = {
+            "update_count": update_count,
+            "row_metadata": row_metadata,
+            "row_page": row_page,
+            "error": error,
+        }
+
+        self.set_future_result_or_exception(response, sql_execute_codec._REQUEST_MESSAGE_TYPE)
+
+    def set_execute_error(self, error):
+        self.set_future_result_or_exception(error, sql_execute_codec._REQUEST_MESSAGE_TYPE)
+
+    def get_message_type(self, invocation):
+        return LE_INT.unpack_from(invocation.request.buf, _OUTBOUND_MESSAGE_MESSAGE_TYPE_OFFSET)[0]
+
+    def set_future_result_or_exception(self, value, message_type):
+        for invocation in self.invocation_registry.values():
+            if self.get_message_type(invocation) == message_type:
+                if isinstance(value, Exception):
+                    invocation.future.set_exception(value)
+                else:
+                    invocation.future.set_result(value)
```

Comment:
I don't like this approach. As a result of `close`, `fetch` calls you get a future, why not directly set_result of it but use invitation registry. I think it is unnecessary complexity.

Reply:
We are mocking things like invocation service, connection manager. etc. to use the code at the internal SQL service. This is a simulation of the server response, if we were to set the result of the future directly, I think that would be cheating :smile:

Reply:
No it would be simpler mocking in my opinion. Mocking is fine in unit testing, adding invocation registry to the table adds complexity to the mocking we try to do.

Reply:
But if you feel this is simpler for you it's ok